cellarium/ml/module/geneformer.py

```diff
@@ -0,0 +1,128 @@
+        tokens = self.feature_ids.expand(x_ng.shape)
+        # sort by median-scaled gene values
+        sorted_indices = torch.argsort(x_ng, dim=1, descending=True)
+        sorted_indices = sorted_indices[:, : self.model.config.max_position_embeddings]
+        ndx = torch.arange(x_ng.shape[0], device=x_ng.device)
+        input_ids = tokens[ndx[:, None], sorted_indices]
+        # mask out genes with zero expression
+        sorted_x_ng = x_ng[ndx[:, None], sorted_indices]
+        attention_mask = sorted_x_ng != 0
+        # pad genes with zero expression
+        input_ids[~attention_mask] = 0
+        return input_ids, attention_mask
+
+    def forward(self, x_ng: torch.Tensor, feature_list: Sequence) -> torch.Tensor:
+        if self.validate_input:
+            assert x_ng.shape[1] == len(feature_list), "The number of x_ng columns must match the feature_list length."
+            assert np.array_equal(feature_list, self.feature_schema), "feature_list must match the feature_schema."
+
+        if self.transform is not None:
+            x_ng = self.transform(x_ng)
+
+        input_ids, attention_mask = self.tokenize(x_ng, feature_list)
+
+        labels = input_ids.clone()
+        labels_probs = torch.full(labels.shape, self.mlm_probability, device=x_ng.device)
+        labels_probs[~attention_mask] = 0
+        masked_indices = torch.bernoulli(labels_probs).bool()
+        labels[~masked_indices] = -100  # we only compute loss on masked tokens
+
+        # 80% of the time, we replace masked input tokens with tokenizer.mask_token ([MASK])
+        indices_replaced = torch.bernoulli(torch.full(labels.shape, 0.8, device=x_ng.device)).bool() & masked_indices
+        input_ids[indices_replaced] = 1  # tokenizer.mask_token_id
+
+        # 10% of the time, we replace masked input tokens with random word
+        indices_random = (
+            torch.bernoulli(torch.full(labels.shape, 0.5, device=x_ng.device)).bool()
+            & masked_indices
+            & ~indices_replaced
+        )
+        random_words = torch.randint(x_ng.shape[1], labels.shape, dtype=torch.long, device=x_ng.device)
+        input_ids[indices_random] = random_words[indices_random]
+
+        output = self.model(
+            input_ids=input_ids,
+            attention_mask=attention_mask,
+            labels=labels,
+        )
+        return output.loss
+
+    def predict(self, x_ng: torch.Tensor, feature_list: Sequence) -> ModelOutput:
+        if self.validate_input:
+            assert x_ng.shape[1] == len(feature_list), "The number of x_ng columns must match the feature_list length."
+            assert np.array_equal(feature_list, self.feature_schema), "feature_list must match the feature_schema."
+
+        if self.transform is not None:
+            x_ng = self.transform(x_ng)
+
+        input_ids, attention_mask = self.tokenize(x_ng, feature_list)
+
+        output = self.model(
```

Comment:
See my earlier comment about the signature of `predict`. Do we _need_ to have the output as `ModelOutput`, or can be cherrypick whatever we need from it and return something more mundane, like `Dict[str, torch.Tensor]`?